pkg/selector/aws/selector.go

```diff
@@ -17,16 +17,149 @@ package aws
+	"github.com/chaos-mesh/chaos-mesh/pkg/mock"
+	"github.com/chaos-mesh/chaos-mesh/pkg/selector/generic"
+	"go.uber.org/fx"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-type SelectImpl struct{}
+// EC2Client defines the minimum client interface required for this package
+type EC2Client interface {
+	DescribeInstances(context.Context, *ec2.DescribeInstancesInput, ...func(*ec2.Options)) (*ec2.DescribeInstancesOutput, error)
+}
+
+type SelectImpl struct {
+	c client.Client
+	generic.Option
+}
 
 func (impl *SelectImpl) Select(ctx context.Context, awsSelector *v1alpha1.AWSSelector) ([]*v1alpha1.AWSSelector, error) {
-	return []*v1alpha1.AWSSelector{awsSelector}, nil
+	if len(awsSelector.Filters) == 0 {
+		return []*v1alpha1.AWSSelector{awsSelector}, nil
+	}
+
+	// we have filters, so we should lookup the cloud resources
+	instances := []*v1alpha1.AWSSelector{}
+
+	ec2client, err := impl.newEc2Client(ctx, awsSelector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create client: %w", err)
+	}
+
+	result, err := ec2client.DescribeInstances(ctx, &ec2.DescribeInstancesInput{
+		Filters: buildEc2Filters(awsSelector.Filters),
+	})
+	if err != nil {
+		return instances, err
+	}
+	for _, r := range result.Reservations {
+		// Set the Ec2Instance, and copy over the other attributes, except the filter
+		instances = append(instances, &v1alpha1.AWSSelector{
+			Ec2Instance: *r.Instances[0].InstanceId,
+			Endpoint:    awsSelector.Endpoint,
+			AWSRegion:   awsSelector.AWSRegion,
+			SecretName:  awsSelector.SecretName,
+			EbsVolume:   awsSelector.EbsVolume,
+			DeviceName:  awsSelector.DeviceName,
+		})
+	}
+	mode := awsSelector.Mode
+	value := awsSelector.Value
+
+	filteredInstances, err := filterInstancesByMode(instances, mode, value)
+	if err != nil {
+		return nil, err
+	}
+
+	return filteredInstances, nil
+}
+
+type Params struct {
+	fx.In
+
+	Client client.Client
+}
+
+func New(params Params) *SelectImpl {
+	return &SelectImpl{
+		params.Client,
+		generic.Option{
+			TargetNamespace: config.ControllerCfg.TargetNamespace,
+		},
+	}
+}
+
+func buildEc2Filters(filters []*v1alpha1.AWSFilter) []ec2types.Filter {
+
+	ec2Filters := []ec2types.Filter{}
+	for _, filter := range filters {
+		ec2Filters = append(ec2Filters, ec2types.Filter{
+			Name:   aws.String(filter.Name),
+			Values: filter.Values,
+		})
+	}
+	return ec2Filters
+}
+
+func (impl *SelectImpl) newEc2Client(ctx context.Context, awsSelector *v1alpha1.AWSSelector) (EC2Client, error) {
+
+	if ec2client := mock.On("MockCreateEc2Client"); ec2client != nil {
+		return ec2client.(EC2Client), nil
+	}
+	opts := []func(*awscfg.LoadOptions) error{
+		awscfg.WithRegion(awsSelector.AWSRegion),
+	}
+
+	if awsSelector.Endpoint != nil {
+		opts = append(opts, awscfg.WithEndpointResolver(aws.EndpointResolverFunc(func(service, region string) (aws.Endpoint, error) {
+			return aws.Endpoint{URL: *awsSelector.Endpoint, SigningRegion: region}, nil
+		})))
+	}
+
+	// TODO How to get namespace for secret??
```

Comment:
I could use some help here.  I can't see how to get the `namespace` in the selector module.

It seems like the pod selector must do this, but I couldn't really follow the code.

Reply:
we assume that the secret is in the same namespace with the `AWSChaos`, so we could use the namespace of AWSChaos.

Is it OK?

Reply:
Yes, it seems to be good :+1: